src/main/scala/shell/Shell.scala

```diff
@@ -59,3 +80,38 @@ abstract class Shell()(implicit p: Parameters) extends LazyModule with LazyScope
   // feel free to override this if necessary
   lazy val module = new LazyRawModuleImp(this)
 }
+
+case class MarkDUTAnnotation(target: ModuleName) extends SingleTargetAnnotation[ModuleName] {
+  def duplicate(n: ModuleName): MarkDUTAnnotation = MarkDUTAnnotation(n)
+}
+
+/** Contains helper methods for finding things relative to the DUT while running other transforms */
+object MarkDUTAnnotation {
+  /** Find names of all [[DefModule]]s in the DUT
+    * Throws exception if no MarkDUTAnnotation found
+    * @return (Name of DUT Top, names of other DefModules in the DUT)
+    */
+  def getDUTModules(state: CircuitState): (String, Set[String]) =
+    getDUTModules(state, new InstanceGraph(state.circuit))
+  def getDUTModules(state: CircuitState, igraph: => InstanceGraph): (String, Set[String]) =
+    getDUTModulesOpt(state, igraph).getOrElse(
+      throw new Exception("Attemping to find all Modules in DUT but no MarkDUTAnnotation found!")
+    )
+
+  /** Find names of all [[DefModule]]s in the DUT if annotation found, None otherwise
+    * @return Option(Name of DUT Top, names of other DefModules in the DUT)
+    */
+  def getDUTModulesOpt(state: CircuitState): Option[(String, Set[String])] =
+    getDUTModulesOpt(state, new InstanceGraph(state.circuit))
+  def getDUTModulesOpt(state: CircuitState, igraph: => InstanceGraph): Option[(String, Set[String])] =
+    state.annotations.collect { case MarkDUTAnnotation(ModuleName(dutTop, _)) => dutTop } match {
+      case Seq() => None
+      case Seq(dutTop) =>
+        val mgraph = igraph.graph.transformNodes(_.module)
+        require(mgraph.contains(dutTop),
+          s"MarkDUTAnnotation found but $dutTop does not exist in circuit!")
+        Some((dutTop, mgraph.reachableFrom(dutTop).toSet))
+      case duts =>
+        throw new Exception(s"Error! More than one DUT specified: " + duts.mkString(", "))
+  }
+}
```

Comment:
Was this code taken from our internal repositories? Do we now have two copies of this?

Reply:
Yeah this is from internal since that stuff is closed source, but we wanted to do it with newshells. I could remove it, but I know Henry filed a Jira for Jack to open source the markdut. Until this happens, we might have to have the 2 copies